# MLM Estudiantes de una Escuela Secundaria.

In [1]:
import pandas as pd
import numpy as np

Sobre el Conjunto de Datos

Contexto:

Los datos fueron obtenidos en una encuesta a estudiantes de cursos de matemáticas y portugués en secundaria. Contiene información social, de género y de estudio interesante sobre los estudiantes. Puedes utilizarlo para realizar un análisis exploratorio de datos (EDA) o intentar predecir la nota final de los estudiantes.

Contenido:

Atributos para ambos conjuntos de datos: student-mat.csv (curso de Matemáticas) y student-por.csv (curso de Portugués):

school: escuela del estudiante (binario: 'GP' - Gabriel Pereira o 'MS' - Mousinho da Silveira)

sex: sexo del estudiante (binario: 'F' - femenino o 'M' - masculino)

age: edad del estudiante (numérico: de 15 a 22)

address: tipo de dirección del hogar del estudiante (binario: 'U' - urbano o 'R' - rural)

famsize: tamaño de la familia (binario: 'LE3' - menor o igual a 3 o 'GT3' - mayor a 3)

Pstatus: estado de convivencia de los padres (binario: 'T' - viviendo juntos o 'A' - separados)

Medu: educación de la madre (numérico: 0 - ninguna, 1 - educación primaria (4º grado), 2 – 5º a 9º grado, 3 – educación secundaria o 4 – educación superior)

Fedu: educación del padre (numérico: 0 - ninguna, 1 - educación primaria (4º grado), 2 – 5º a 9º grado, 3 – educación secundaria o 4 – educación superior)

Mjob: ocupación de la madre (nominal: 'teacher' - profesora, 'health' - relacionado con la salud, 'services' - servicios civiles (por ejemplo, administrativa o policía), 'at_home' - en casa u 'other' - otro)

Fjob: ocupación del padre (nominal: 'teacher' - profesor, 'health' - relacionado con la salud, 'services' - servicios civiles (por ejemplo, administrativa o policía), 'at_home' - en casa u 'other' - otro)

reason: motivo para elegir esta escuela (nominal: cerca de 'home' - casa, reputación de la 'school' - escuela, preferencia por el 'course' - curso u 'other' - otro)

guardian: tutor del estudiante (nominal: 'mother' - madre, 'father' - padre u 'other' - otro)

traveltime: tiempo de viaje de casa a la escuela (numérico: 1 - <15 min., 2 - 15 a 30 min., 3 - 30 min. a 1 hora, o 4 - >1 hora)

studytime: tiempo de estudio semanal (numérico: 1 - <2 horas, 2 - 2 a 5 horas, 3 - 5 a 10 horas, o 4 - >10 horas)

failures: número de fracasos en clases anteriores (numérico: n si 1<=n<3, de lo contrario 4)

schoolsup: apoyo educativo adicional (binario: sí o no)

famsup: apoyo educativo familiar (binario: sí o no)

paid: clases adicionales pagadas dentro del curso (Matemáticas o Portugués) (binario: sí o no)

activities: actividades extracurriculares (binario: sí o no)

nursery: asistencia a la escuela infantil (binario: sí o no)

higher: desea continuar estudios superiores (binario: sí o no)

internet: acceso a Internet en casa (binario: sí o no)

romantic: con relación romántica (binario: sí o no)

famrel: calidad de las relaciones familiares (numérico: de 1 - muy mala a 5 - excelente)

freetime: tiempo libre después de la escuela (numérico: de 1 - muy bajo a 5 - muy alto)

goout: salir con amigos (numérico: de 1 - muy bajo a 5 - muy alto)

Dalc: consumo de alcohol en días laborales (numérico: de 1 - muy bajo a 5 - muy alto)

Walc: consumo de alcohol durante el fin de semana (numérico: de 1 - muy bajo a 5 - muy alto)

health: estado de salud actual (numérico: de 1 - muy malo a 5 - muy bueno)

absences: número de ausencias escolares (numérico: de 0 a 93)

Estas calificaciones están relacionadas con el curso, Matemáticas o Portugués:

G1: nota del primer periodo (numérico: de 0 a 20)

G2: nota del segundo periodo (numérico: de 0 a 20)

G3: nota final (numérico: de 0 a 20, objetivo de salida)

In [13]:
urlmath='/home/julio/jupyter_files/MLM_StudentAlcoholConsumption/student-mat.csv'
# Leer el archivo CSV
df1 = pd.read_csv(urlmath)
df1['course'] = 'Math'

In [14]:
# Mostrar las primeras filas del DataFrame
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [15]:
urlpor='/home/julio/jupyter_files/MLM_StudentAlcoholConsumption/student-por.csv'
# Leer el archivo CSV
df2 = pd.read_csv(urlpor)
df2['course'] = 'Por'

In [16]:
# Mostrar las primeras filas del DataFrame
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      649 non-null    object
 1   sex         649 non-null    object
 2   age         649 non-null    int64 
 3   address     649 non-null    object
 4   famsize     649 non-null    object
 5   Pstatus     649 non-null    object
 6   Medu        649 non-null    int64 
 7   Fedu        649 non-null    int64 
 8   Mjob        649 non-null    object
 9   Fjob        649 non-null    object
 10  reason      649 non-null    object
 11  guardian    649 non-null    object
 12  traveltime  649 non-null    int64 
 13  studytime   649 non-null    int64 
 14  failures    649 non-null    int64 
 15  schoolsup   649 non-null    object
 16  famsup      649 non-null    object
 17  paid        649 non-null    object
 18  activities  649 non-null    object
 19  nursery     649 non-null    object
 20  higher    

In [18]:
df3 = pd.concat([df1, df2], axis=0, ignore_index=True)

In [83]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1439 entries, 0 to 1438
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      1439 non-null   object
 1   sex         1439 non-null   object
 2   age         1439 non-null   int64 
 3   address     1439 non-null   object
 4   famsize     1439 non-null   object
 5   Pstatus     1439 non-null   object
 6   Medu        1439 non-null   int64 
 7   Fedu        1439 non-null   int64 
 8   Mjob        1439 non-null   object
 9   Fjob        1439 non-null   object
 10  reason      1439 non-null   object
 11  guardian    1439 non-null   object
 12  traveltime  1439 non-null   int64 
 13  studytime   1439 non-null   int64 
 14  failures    1439 non-null   int64 
 15  schoolsup   1439 non-null   object
 16  famsup      1439 non-null   object
 17  paid        1439 non-null   object
 18  activities  1439 non-null   object
 19  nursery     1439 non-null   object
 20  higher  

In [20]:
# Calcular el porcentaje de valores nulos por columna
null_percentage = df3.isnull().mean() * 100
# Mostrar el porcentaje de valores nulos
print(null_percentage)

school        0.0
sex           0.0
age           0.0
address       0.0
famsize       0.0
Pstatus       0.0
Medu          0.0
Fedu          0.0
Mjob          0.0
Fjob          0.0
reason        0.0
guardian      0.0
traveltime    0.0
studytime     0.0
failures      0.0
schoolsup     0.0
famsup        0.0
paid          0.0
activities    0.0
nursery       0.0
higher        0.0
internet      0.0
romantic      0.0
famrel        0.0
freetime      0.0
goout         0.0
Dalc          0.0
Walc          0.0
health        0.0
absences      0.0
G1            0.0
G2            0.0
G3            0.0
course        0.0
dtype: float64


**Preprocesar Datos**

In [84]:
# Importar librerías
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from imblearn.over_sampling import SMOTE

In [87]:
# Separar características y variable objetivo
X = df3.drop('G3', axis=1)
y = df3['G3']

# Identificar columnas categóricas y numéricas
categorical_cols = X.select_dtypes(include=['object']).columns
numerical_cols = X.select_dtypes(include=['int', 'float']).columns

In [88]:
# Definir preprocesador
preprocessor = ColumnTransformer(
    transformers=[
        ('num', Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='mean')),
            ('scaler', StandardScaler())
        ]), numerical_cols),
        ('cat', Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('onehot', OneHotEncoder(drop='first'))
        ]), categorical_cols)
    ]
)

In [89]:
# Dividir datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [43]:
# Aplicar SMOTE para balanceo de clases en el conjunto de entrenamiento
#smote = SMOTE(random_state=42)
#X_train_balanced, y_train_balanced = smote.fit_resample(X_train, y_train)

**Modelo de Regresión Logística**

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

# Crear pipeline para la regresión logística
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000))
])

# Ajustar el modelo
model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2'],
      dtype='object')),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehot',
                                                                   OneHotEncoder(drop='first'))]),
                                                  Index(['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob',
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'course'],
      dtype='object'))])),
                ('classifier', LogisticRegression(max_iter=1000))])

In [27]:
# Evaluar el modelo
y_pred = model.predict(X_test)
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(f'Classification Report:\n{classification_report(y_test, y_pred)}')

Accuracy: 0.375
Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.65      0.69        31
           1       0.00      0.00      0.00         0
           5       0.40      1.00      0.57         2
           6       0.50      0.27      0.35        11
           7       1.00      0.14      0.25         7
           8       0.41      0.43      0.42        28
           9       0.29      0.06      0.10        35
          10       0.35      0.63      0.45        59
          11       0.30      0.40      0.34        52
          12       0.32      0.16      0.22        43
          13       0.38      0.35      0.37        43
          14       0.30      0.19      0.23        42
          15       0.35      0.52      0.41        33
          16       0.27      0.20      0.23        20
          17       0.40      0.46      0.43        13
          18       0.45      0.50      0.48        10
          19       0.67      0.67      0.6

/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Optimización del Modelo**

In [30]:
from sklearn.model_selection import GridSearchCV

# Definir parámetros para la búsqueda de hiperparámetros
param_grid = {
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__solver': ['liblinear', 'lbfgs']
}

In [31]:
# Crear y ajustar el modelo con búsqueda de hiperparámetros
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy')

In [32]:
grid_search.fit(X_train, y_train)

/home/julio/.local/lib/python3.10/site-packages/sklearn/model_selection/_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2'],
      dtype='object')),
                                                                        ('cat',
                                                                         Pipeline(steps=[(...
                                                                         Index(['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob',
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'course'],
      dtype='object'))])),
                                       ('classifier',
                                        LogisticRegression(max_iter=1000))]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__solver': ['liblinear', 'lbfgs']},
             scoring='accuracy')

In [34]:
# Resultados de la búsqueda
print(f'Best parameters: {grid_search.best_params_}')
print(f'Best score: {grid_search.best_score_}')


Best parameters: {'classifier__C': 100, 'classifier__solver': 'lbfgs'}
Best score: 0.5004876607063691


In [35]:
# Evaluar el modelo optimizado
best_model = grid_search.best_estimator_
y_pred_optimized = best_model.predict(X_test)
print(f'Optimized Accuracy: {accuracy_score(y_test, y_pred_optimized)}')
print(f'Optimized Classification Report:\n{classification_report(y_test, y_pred_optimized)}')

Optimized Accuracy: 0.48148148148148145
Optimized Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.71      0.75        31
           5       0.33      1.00      0.50         2
           6       0.50      0.36      0.42        11
           7       0.50      0.71      0.59         7
           8       0.49      0.64      0.55        28
           9       0.64      0.26      0.37        35
          10       0.49      0.63      0.55        59
          11       0.36      0.40      0.38        52
          12       0.38      0.26      0.31        43
          13       0.44      0.47      0.45        43
          14       0.44      0.40      0.42        42
          15       0.59      0.61      0.60        33
          16       0.38      0.30      0.33        20
          17       0.44      0.62      0.52        13
          18       0.55      0.60      0.57        10
          19       0.67      0.67      0.67         3

    acc

**Modelo de SVM:Crear y Ajustar.** 

In [44]:
from sklearn.svm import SVC

# Crear pipeline para el modelo SVM
model_svm = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', SVC())
])

# Ajustar el modelo
model_svm.fit(X_train, y_train)

# Evaluar el modelo
y_pred_svm = model_svm.predict(X_test)
print(f'Accuracy: {accuracy_score(y_test, y_pred_svm)}')
print(f'Classification Report:\n{classification_report(y_test, y_pred_svm)}')

Accuracy: 0.41435185185185186
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.65      0.71        31
           5       0.00      0.00      0.00         2
           6       1.00      0.27      0.43        11
           7       0.00      0.00      0.00         7
           8       0.50      0.36      0.42        28
           9       0.75      0.09      0.15        35
          10       0.35      0.75      0.48        59
          11       0.36      0.62      0.45        52
          12       0.32      0.14      0.19        43
          13       0.43      0.37      0.40        43
          14       0.35      0.26      0.30        42
          15       0.38      0.73      0.49        33
          16       0.67      0.10      0.17        20
          17       0.50      0.08      0.13        13
          18       0.86      0.60      0.71        10
          19       0.50      0.33      0.40         3

    accuracy               

/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Optimización del Modelo**

In [45]:
from sklearn.model_selection import GridSearchCV

# Definir parámetros para la búsqueda de hiperparámetros
param_grid_svm = {
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__kernel': ['linear', 'rbf'],
    'classifier__gamma': ['scale', 'auto']
}

In [46]:
# Crear y ajustar el modelo con búsqueda de hiperparámetros
grid_search_svm = GridSearchCV(model_svm, param_grid_svm, cv=5, scoring='accuracy')

In [47]:
grid_search_svm.fit(X_train, y_train)

/home/julio/.local/lib/python3.10/site-packages/sklearn/model_selection/_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2'],
      dtype='object')),
                                                                        ('cat',
                                                                         Pipeline(steps=[(...
                                                                         Index(['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob',
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'course'],
      dtype='object'))])),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': ['scale', 'auto'],
                         'classifier__kernel': ['linear', 'rbf']},
             scoring='accuracy')

In [48]:
# Resultados de la búsqueda
print(f'Best parameters: {grid_search_svm.best_params_}')
print(f'Best score: {grid_search_svm.best_score_}')

Best parameters: {'classifier__C': 100, 'classifier__gamma': 'auto', 'classifier__kernel': 'rbf'}
Best score: 0.5421801881680705


In [49]:
# Evaluar el modelo optimizado
best_model_svm = grid_search_svm.best_estimator_
y_pred_svm_optimized = best_model_svm.predict(X_test)

In [50]:
print(f'Optimized Accuracy: {accuracy_score(y_test, y_pred_svm_optimized)}')
print(f'Optimized Classification Report:\n{classification_report(y_test, y_pred_svm_optimized)}')

Optimized Accuracy: 0.5717592592592593
Optimized Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.58      0.64        31
           5       1.00      1.00      1.00         2
           6       1.00      0.36      0.53        11
           7       1.00      0.86      0.92         7
           8       0.49      0.75      0.59        28
           9       0.67      0.51      0.58        35
          10       0.59      0.64      0.62        59
          11       0.46      0.56      0.50        52
          12       0.57      0.60      0.58        43
          13       0.53      0.47      0.49        43
          14       0.50      0.45      0.47        42
          15       0.58      0.67      0.62        33
          16       0.67      0.40      0.50        20
          17       0.58      0.54      0.56        13
          18       0.67      0.60      0.63        10
          19       0.60      1.00      0.75         3

    accu

**Modelo KNN**

In [94]:
# Crear el modelo KNN
knn = KNeighborsClassifier()

# Crear el pipeline que incluye el preprocesador y el modelo
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', knn)
])

# Ajustar el modelo
pipeline.fit(X_train, y_train)

# Predecir en los datos de prueba
y_pred = pipeline.predict(X_test)

In [95]:
# Evaluar el modelo
accuracy_before = accuracy_score(y_test, y_pred)
classification_rep_before = classification_report(y_test, y_pred)
conf_matrix_before = confusion_matrix(y_test, y_pred)

print("Before Optimization:")
print(f"Accuracy: {accuracy_before}")
print("\nClassification Report:")
print(classification_rep_before)
print("Confusion Matrix:")
print(conf_matrix_before)

Before Optimization:
Accuracy: 0.2777777777777778

Classification Report:
              precision    recall  f1-score   support

           0       0.39      0.52      0.44        31
           5       0.00      0.00      0.00         2
           6       0.50      0.36      0.42        11
           7       0.12      0.14      0.13         7
           8       0.19      0.25      0.22        28
           9       0.04      0.03      0.03        35
          10       0.27      0.34      0.30        59
          11       0.24      0.33      0.28        52
          12       0.38      0.30      0.34        43
          13       0.30      0.23      0.26        43
          14       0.26      0.24      0.25        42
          15       0.31      0.27      0.29        33
          16       0.36      0.25      0.29        20
          17       0.17      0.08      0.11        13
          18       0.62      0.50      0.56        10
          19       0.50      0.33      0.40         3

    ac

**Optimización del modelo con GridSearchCV**

In [96]:
from sklearn.model_selection import GridSearchCV

# Definir los parámetros para la búsqueda en cuadrícula (GridSearch)
param_grid = {
    'classifier__n_neighbors': [3, 5, 7, 10],
    'classifier__weights': ['uniform', 'distance'],
    'classifier__p': [1, 2]  # p=1 para Manhattan, p=2 para Euclidean
}

# Crear la búsqueda en cuadrícula
grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2)

# Ajustar el modelo
grid_search.fit(X_train, y_train)


Fitting 5 folds for each of 16 candidates, totalling 80 fits


/home/julio/.local/lib/python3.10/site-packages/sklearn/model_selection/_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2'],
      dtype='object')),
                                                                        ('cat',
                                                                         Pipeline(steps=[(...
                                                                         Index(['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob',
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'course'],
      dtype='object'))])),
                                       ('classifier', KNeighborsClassifier())]),
             n_jobs=-1,
             param_grid={'classifier__n_neighbors': [3, 5, 7, 10],
                         'classifier__p': [1, 2],
                         'classifier__weights': ['uniform', 'distance']},
             verbose=2)

In [97]:
# Obtener el mejor modelo
best_knn = grid_search.best_estimator_

# Predecir en los datos de prueba con el mejor modelo
y_pred_optimized = best_knn.predict(X_test)

# Evaluar el modelo optimizado
accuracy_after = accuracy_score(y_test, y_pred_optimized)
classification_rep_after = classification_report(y_test, y_pred_optimized)
conf_matrix_after = confusion_matrix(y_test, y_pred_optimized)

In [98]:
print("After Optimization:")
print(f"Accuracy: {accuracy_after}")
print("\nClassification Report:")
print(classification_rep_after)
print("Confusion Matrix:")
print(conf_matrix_after)

After Optimization:
Accuracy: 0.4837962962962963

Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.52      0.58        31
           5       0.50      1.00      0.67         2
           6       0.67      0.36      0.47        11
           7       1.00      0.71      0.83         7
           8       0.59      0.61      0.60        28
           9       0.52      0.34      0.41        35
          10       0.40      0.53      0.46        59
          11       0.41      0.58      0.48        52
          12       0.57      0.49      0.53        43
          13       0.51      0.44      0.47        43
          14       0.39      0.36      0.38        42
          15       0.38      0.52      0.44        33
          16       0.67      0.30      0.41        20
          17       0.50      0.38      0.43        13
          18       0.67      0.60      0.63        10
          19       0.50      1.00      0.67         3

    acc

In [99]:
# Imprimir los mejores parámetros y el mejor puntaje
print("Best Parameters Found:")
print(grid_search.best_params_)
print(f"Best Accuracy from Grid Search: {grid_search.best_score_}")

Best Parameters Found:
{'classifier__n_neighbors': 10, 'classifier__p': 1, 'classifier__weights': 'distance'}
Best Accuracy from Grid Search: 0.484611595487907


**Modelo GBM**

In [101]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [102]:
# Crear el modelo GBM
gbm = GradientBoostingClassifier()

# Crear el pipeline que incluye el preprocesador y el modelo
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', gbm)
])

# Ajustar el modelo
pipeline.fit(X_train, y_train)

# Predecir en los datos de prueba
y_pred = pipeline.predict(X_test)


In [103]:
# Evaluar el modelo
accuracy_before = accuracy_score(y_test, y_pred)
classification_rep_before = classification_report(y_test, y_pred)
conf_matrix_before = confusion_matrix(y_test, y_pred)

print("Before Optimization:")
print(f"Accuracy: {accuracy_before}")
print("\nClassification Report:")
print(classification_rep_before)
print("Confusion Matrix:")
print(conf_matrix_before)

Before Optimization:
Accuracy: 0.6296296296296297

Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.74      0.77        31
           1       0.00      0.00      0.00         0
           5       0.67      1.00      0.80         2
           6       1.00      0.55      0.71        11
           7       1.00      0.71      0.83         7
           8       0.56      0.79      0.66        28
           9       0.54      0.37      0.44        35
          10       0.53      0.66      0.59        59
          11       0.59      0.62      0.60        52
          12       0.60      0.56      0.58        43
          13       0.60      0.70      0.65        43
          14       0.64      0.50      0.56        42
          15       0.71      0.73      0.72        33
          16       0.80      0.60      0.69        20
          17       0.78      0.54      0.64        13
          18       0.75      0.90      0.82        10
       

/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


[CV] END classifier__n_neighbors=3, classifier__p=1, classifier__weights=uniform; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=1, classifier__weights=distance; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=1, classifier__weights=distance; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=uniform; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=uniform; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=uniform; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=distance; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=distance; total time=   0.1s
[CV] END classifier__n_neighbors=3, classifier__p=2, classifier__weights=distance; total time=   0.1s
[CV] END classifier__n_neighbors=5, classifier__p=1, classifier__weights=uniform; tota

**Optimización del Modelo con Grid Search**

In [105]:
from sklearn.model_selection import GridSearchCV

# Definir los parámetros simplificados para la búsqueda en cuadrícula (GridSearch)
param_grid = {
    'classifier__n_estimators': [100, 200],  # Menos valores
    'classifier__learning_rate': [0.1],       # Menos valores
    'classifier__max_depth': [3, 5],          # Menos valores
    'classifier__min_samples_split': [2, 5]   # Menos valores
}

# Crear la búsqueda en cuadrícula
grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2)

# Ajustar el modelo
grid_search.fit(X_train, y_train)

Fitting 5 folds for each of 8 candidates, totalling 40 fits


/home/julio/.local/lib/python3.10/site-packages/sklearn/model_selection/_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2'],
      dtype='object')),
                                                                        ('cat',
                                                                         Pipeline(steps=[(...
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'course'],
      dtype='object'))])),
                                       ('classifier',
                                        GradientBoostingClassifier())]),
             n_jobs=-1,
             param_grid={'classifier__learning_rate': [0.1],
                         'classifier__max_depth': [3, 5],
                         'classifier__min_samples_split': [2, 5],
                         'classifier__n_estimators': [100, 200]},
             verbose=2)

In [106]:
# Obtener el mejor modelo
best_gbm = grid_search.best_estimator_

In [107]:
# Predecir en los datos de prueba con el mejor modelo
y_pred_optimized = best_gbm.predict(X_test)

In [108]:
# Evaluar el modelo optimizado
accuracy_after = accuracy_score(y_test, y_pred_optimized)
classification_rep_after = classification_report(y_test, y_pred_optimized)
conf_matrix_after = confusion_matrix(y_test, y_pred_optimized)

/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/julio/.local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1509: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [109]:
print("After Optimization:")
print(f"Accuracy: {accuracy_after}")
print("\nClassification Report:")
print(classification_rep_after)
print("Confusion Matrix:")
print(conf_matrix_after)

After Optimization:
Accuracy: 0.6226851851851852

Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.61      0.70        31
           1       0.00      0.00      0.00         0
           5       1.00      1.00      1.00         2
           6       1.00      0.55      0.71        11
           7       1.00      0.71      0.83         7
           8       0.51      0.79      0.62        28
           9       0.54      0.37      0.44        35
          10       0.49      0.68      0.57        59
          11       0.65      0.60      0.62        52
          12       0.63      0.60      0.62        43
          13       0.60      0.70      0.65        43
          14       0.62      0.50      0.55        42
          15       0.69      0.73      0.71        33
          16       0.86      0.60      0.71        20
          17       0.70      0.54      0.61        13
          18       0.73      0.80      0.76        10
        

In [110]:
# Imprimir los mejores parámetros y el mejor puntaje
print("Best Parameters Found:")
print(grid_search.best_params_)
print(f"Best Accuracy from Grid Search: {grid_search.best_score_}")


Best Parameters Found:
{'classifier__learning_rate': 0.1, 'classifier__max_depth': 3, 'classifier__min_samples_split': 5, 'classifier__n_estimators': 100}
Best Accuracy from Grid Search: 0.6276242549628097
[CV] END classifier__learning_rate=0.1, classifier__max_depth=3, classifier__min_samples_split=2, classifier__n_estimators=100; total time=   9.1s
[CV] END classifier__learning_rate=0.1, classifier__max_depth=3, classifier__min_samples_split=2, classifier__n_estimators=200; total time=  15.9s
[CV] END classifier__learning_rate=0.1, classifier__max_depth=3, classifier__min_samples_split=2, classifier__n_estimators=200; total time=  16.8s
[CV] END classifier__learning_rate=0.1, classifier__max_depth=3, classifier__min_samples_split=5, classifier__n_estimators=100; total time=   8.4s
[CV] END classifier__learning_rate=0.1, classifier__max_depth=3, classifier__min_samples_split=5, classifier__n_estimators=200; total time=  16.1s
[CV] END classifier__learning_rate=0.1, classifier__max_dep

**Resultados Detallados**

1. **Gradient Boosting Machines (GBM)**
   - **Accuracy:** 0.62 → 0.62 (sin cambio significativo)
   - **Precision (Macro Avg):** 0.67 → 0.68 (+0.01)
   - **Recall (Macro Avg):** 0.64 → 0.63 (-0.01)
   - **F1-Score (Macro Avg):** 0.64 → 0.64 (sin cambio significativo)

2. **SVM**
   - **Accuracy:** 0.41 → 0.57 (+0.16)
   - **Precision (Macro Avg):** 0.48 → 0.66 (+0.18)
   - **Recall (Macro Avg):** 0.33 → 0.62 (+0.29)
   - **F1-Score (Macro Avg):** 0.34 → 0.63 (+0.29)

3. **Regresión Logística**
   - **Accuracy:** 0.38 → 0.48 (+0.10)
   - **Precision (Macro Avg):** 0.42 → 0.50 (+0.08)
   - **Recall (Macro Avg):** 0.39 → 0.54 (+0.15)
   - **F1-Score (Macro Avg):** 0.36 → 0.50 (+0.14)

4. **KNN**
   - **Accuracy:** 0.28 → 0.28 (sin cambio)
   - **Precision (Macro Avg):** 0.29 → 0.29 (sin cambio)
   - **Recall (Macro Avg):** 0.26 → 0.26 (sin cambio)
   - **F1-Score (Macro Avg):** 0.27 → 0.27 (sin cambio)

### Conclusión

- **Gradient Boosting Machines (GBM)** sigue siendo el modelo más eficaz en términos de **Accuracy** y **F1-Score** con valores estables, mostrando un buen rendimiento general.
  
- **SVM** ha mostrado una mejora significativa en todas las métricas, especialmente en **Recall** y **F1-Score**, destacándose notablemente después de la optimización. Sin embargo, su **Accuracy** es aún inferior al de GBM.

- **Regresión Logística** también ha mejorado, pero sus métricas siguen siendo inferiores a las de GBM y SVM, lo que sugiere que aunque ha tenido avances, no es el modelo más competitivo.

- **KNN** no ha mostrado mejoras, manteniéndose como el modelo menos efectivo en esta comparación.

En resumen, **GBM** es el mejor modelo en términos generales, mientras que **SVM** ha mejorado considerablemente pero aún no supera a GBM. **Regresión Logística** ha mostrado mejoras, y **KNN** sigue siendo el menos eficaz.